crates/dbs-miniball/docs/MemoryVirtualization.md

```diff
@@ -0,0 +1,381 @@
+    loop {
+        match self.event_mgr.run() {
+            Ok(_) => (),
+            Err(e) => eprintln!("Failed to handle events: {:?}", e),
+        }
+        if !self.exit_handler.keep_running() {
+            break;
+        }
+    }
+    self.vm.shutdown();
+
+    Ok(())
+    }
+```
+
+加载内核的过程如下：
+
+1. 加载内核镜像到 guest 内存中
+1. 加载内核命令行到 guest 内存中
+1. 配置启动参数
+
+#### 加载内核镜像
+第 4 行，加载了 kernel 的镜像。第 6 ～ 24 行，通过 ELF 将 kernel 的镜像加载到 Guest 内存中。
+
+```rust
+// Load the kernel into guest memory.
+#[cfg(target_arch = "x86_64")]
+fn load_kernel(&mut self) -> Result<KernelLoaderResult> {
+    let mut kernel_image = File::open(&self.kernel_cfg.path).map_err(Error::IO)?;
+
+    // Load the kernel into guest memory.
+    let kernel_load = match Elf::load(
+        &self.guest_memory,
+        None,
+        &mut kernel_image,
+        Some(GuestAddress(self.kernel_cfg.load_addr)),
+    ) {
+        Ok(result) => result,
+        Err(loader::Error::Elf(elf::Error::InvalidElfMagicNumber)) => BzImage::load(
+            &self.guest_memory,
+            None,
+            &mut kernel_image,
+            Some(GuestAddress(self.kernel_cfg.load_addr)),
+        )
+        .map_err(Error::KernelLoad)?,
+        Err(e) => {
+            return Err(Error::KernelLoad(e));
+        }
+    };
+
+    ...
+    
+    Ok(kernel_load)
+}
+```
+
+#### 加载内核命令行
+第 6 ～ 13 行，加载内核命令行到 Guest 内存中，传入 Guest memory 和命令行起始地址，并保留命令行地址待之后输入到启动参数中。
+
+```rust
+// Load the kernel into guest memory.
+#[cfg(target_arch = "x86_64")]
+fn load_kernel(&mut self) -> Result<KernelLoaderResult> {
+    ...
+
+    // Load the kernel command line into guest memory.
+    let cmdline_addr = dbs_boot::layout::CMDLINE_START;
+    linux_loader::loader::load_cmdline(
+        &self.guest_memory,
+        GuestAddress(cmdline_addr),
+        &self.kernel_cfg.cmdline,
+    )
+    .map_err(Error::LoadCommandline)?;
+    
+    ...
+}
+```
+
+#### 配置启动参数
+第 6 ～ 16 行，配置启动参数，参数的详细内容在之后代码中介绍。
+```rust
+// Load the kernel into guest memory.
+#[cfg(target_arch = "x86_64")]
+fn load_kernel(&mut self) -> Result<KernelLoaderResult> {
+    ...
+    
+    // Generate boot parameters.
+    build_bootparams(
+        &self.guest_memory,
+        &self.address_space,
+        GuestAddress(self.kernel_cfg.load_addr),
+        GuestAddress(MMIO_GAP_START),
+        GuestAddress(MMIO_GAP_END),
+        GuestAddress(cmdline_addr),
+        (self.kernel_cfg.cmdline.as_str().len() + 1) as usize,
+    )
+    .map_err(Error::BootParam)?;
+
+    ...
+}
+```
+
+第 9 ～ 11 行，先判断输入的MMIO 起始地址是否小于终止地址。第 13 ～ 22 行，构造启动参数，其中第 20 ～ 22 行，添加内核命令行地址到启动参数中。第 24 ～ 26 行，通过 `add_e820_entry()` 方法将启动参数信息加载到数据区 EBDA 中。第 28 ～ 69 行，根据 Guest 内存的大小与 MMIO 地址范围的关系将启动参数写入对应的位置。
+第 72 ～ 79 行，将启动参数写入 zeropage 中。
+> zero page是一个特殊的物理页，里面值全部为0，zero page是针对匿名页场景专门进行优化，主要是节省内存和对性能进行了一定优化。当malloc或者mapp一段虚拟内存后，第一次对该内存访问为读操作，将会发生匿名page fault。do_anonymous_page处理，由于第一次为读操作还未发生写操作，因此发生一个zero page，为其申请一个特殊物理页zero page。
```

Comment:
The zero page in your code should be referred to this concept (https://www.kernel.org/doc/html/v5.8/x86/zero-page.html) in linux kernel.

Reply:
I re-described it with reference to the Linux kernel docs.